src/main/java/io/jenkins/plugins/opentelemetry/backend/elastic/ElasticsearchLogStorageScrollingRetriever.java

```diff
@@ -0,0 +1,224 @@
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.security.Principal;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+/**
+ * Retrieve the logs from Elasticsearch.
+ * FIXME graceful shutdown
+ */
+public class ElasticsearchLogStorageScrollingRetriever implements LogStorageRetriever {
+    public static final String TIMESTAMP = "@timestamp";
+    public static final Time POINT_IN_TIME_TTL = Time.of(builder -> builder.time("30s"));
+    public static final int PAGE_SIZE = 100; // FIXME
+
+    private final static Logger logger = Logger.getLogger(ElasticsearchLogStorageScrollingRetriever.class.getName());
+
+    @Nonnull
+    private final String indexPattern;
+
+    @Nonnull
+    final transient ElasticsearchClient elasticsearchClient;
+
+
+    /**
+     * TODO verify unsername:password auth vs apiKey auth
+     */
+    public ElasticsearchLogStorageScrollingRetriever(String elasticsearchUrl, Credentials elasticsearchCredentials, String indexPattern) {
+        if (StringUtils.isBlank(elasticsearchUrl)) {
+            throw new IllegalArgumentException("Elasticsearch url cannot be blank");
+        }
+        if (StringUtils.isBlank(indexPattern)) {
+            throw new IllegalArgumentException("Elasticsearch Index Pattern cannot be blank");
+        }
+        this.indexPattern = indexPattern;
+
+        BasicCredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+        credentialsProvider.setCredentials(AuthScope.ANY, elasticsearchCredentials);
+
+        RestClient restClient = RestClient.builder(HttpHost.create(elasticsearchUrl))
+            .setHttpClientConfigCallback(httpClientBuilder -> httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider))
+            .build();
+        RestClientTransport elasticsearchTransport = new RestClientTransport(restClient, new JacksonJsonpMapper());
+        this.elasticsearchClient = new ElasticsearchClient(elasticsearchTransport);
+    }
+
+    @Nonnull
+    @Override
+    public LogsQueryResult overallLog(@Nonnull String traceId, @Nonnull String spanId, @Nullable LogsQueryContext logsQueryContext) throws IOException {
+        // https://www.elastic.co/guide/en/elasticsearch/reference/7.17/scroll-api.html
+
+        Charset charset = StandardCharsets.UTF_8;
+        boolean complete;
+        String newScrollId;
+        List<Hit<ObjectNode>> hits;
+        if (logsQueryContext == null) {
+            SearchRequest searchRequest = new SearchRequest.Builder()
+                .scroll(builder -> builder.time("30s"))
+                .size(PAGE_SIZE)
+                .sort(sortBuilder -> sortBuilder.field(fieldBuilder -> fieldBuilder.field(TIMESTAMP).order(SortOrder.Asc)))
```

Comment:
@swallez are there java constants for standard field names like `@timestamp`?

Reply:
Nope. It could however be part of a separate library to model ECS types.